### ISSUE TITLE
Serialize desktop state to file

### ISSUE THREAD

**Issue by carymrobbins** [koekeishiya/chunkwm#190]
I like having each desktop set up a certain way. On one monitor I have the following setup -

```
[ Gitter   ] [ Slack ]
[ Limechat ] [ Email ]
```

However, I have to move things around manually each time. Ideally, I should be able to use chunkc's serialization mechanism to "remember" the state of the app positions.

It seems I can serialize/deserialize to a file with something like this in my `.khdrc`

```
lalt + lcmd + rcmd - r : chunkc tiling::desktop --deserialize ~/.chunk-desktop-r
lalt + lcmd + rcmd + shift - r : chunkc tiling::desktop --serialize ~/.chunk-desktop-r
```

I can see the file which was serialized, but when I _deserialize_ nothing seems to happen. Also, the serialized file looks something like -

```
% cat ~/.chunk-desktop-r
root vertical 0.500
left_root horizontal 0.500
left_leaf
right_leaf
right_root horizontal 0.500
left_leaf
right_leaf
```

This leads me to believe that it's not actually capturing which application goes where.

Is what I'm trying to accomplish possible with chunkwm?

**Comment by koekeishiya** [comment]
The serialization system can only manage the layout. 

Every window of an application has a unique window id, and this id changes every time you close or open the main window of the application, or the application is restarted. Because of this, serializing with windows is not an option.

I also consider serializing the application to be ambiguous. What happens in cases where there are multiple instances of the same application. 

What if this application is not currently running, do we just leave an empty spot? We'd probably not do that, and tile other applications there, but then what happens when said application is then opened afterwards, and won't tile properly into the layout.

**Comment by dominiklohmann** [comment]
Couldn't this be solved with another command that changes the order windows are stored in internally (e.g. sorts the bsp tree by application name and optional window name/regex)?

Possible syntax:
```
chunkc tiling::desktop --sort-windows [owner=Gitter] [owner=Slack]
chunkc tiling::desktop --sort-windows [owner=Limechat] [owner=Tweetbot name="^Main Window$"] [owner=Mail]
```

If a window/application cannot be found it just gets ignored. Not sure what to do if there are multiple windows matching (sort first, sort all?), but that's what window name regexes can be used for. If there are more windows than specified for the sort, leave the rest at the rest of the tree in order (stable sort).

Sorting just before deserializing should do the trick.

**Comment by koekeishiya** [comment]
@dominiklohmann 

You'd still have to be able to visualize what the tree looks like to be able to get any benefit out of doing it that way, as there are several tree-representations for any given layout. A specific ordering of window would yield different on-screen placements depending on the tree-representation.

We currently use `getopt` to do command parsing. We would need to write a custom parser if we want to do multiple arguments per switch.

I personally don't find the above work to be worth it for this feature, considering it would only be applicable to simple trees, in which it is at most ~3 swap commands.

I might take a look at this if I feel bored and want to kill some time in the future.

**Comment by dominiklohmann** [comment]
> You'd still have to be able to visualize what the tree looks like to be able to get any benefit out of doing it that way, as there are several tree-representations for any given layout. A specific ordering of window would yield different on-screen placements depending on the tree-representation.

Non-issue if used with `--deserialize`, I think. Not sure what other scenario this would ever be used in.

> We currently use getopt to do command parsing. We would need to write a custom parser if we want to do multiple arguments per switch.

Could just change the syntax to require multiple `--sort-windows` in order, one for each window. That's possible with `getopt`.

```
chunkc tiling::desktop --sort-windows "Gitter" --sort-windows "Tweetbot" "^Main Window$"
```

**Comment by koekeishiya** [comment]
I'm probably going to try to do something simple with *swallow* criterias here. 

Basically every leaf node will be serialized with the name of the application and window title. When a deserialization happens, if there is no window that matches the criteria, a placeholder window would be created and would remain there until a real window that matches said criteria is opened.

Maybe being able to have the critera drawn as text in the middle of the placeholder window or something as well could be a nice addition.

**Comment by koekeishiya** [comment]
Made some changes to the serialization system on the branch *serialize-state*. I would appreciate it if you could test this and give some feedback.

The system is very simple, and also serializes the name of the application and the window title to file.
When a deserialization occur, only windows with the exact same application name and window title is able to fill that region. It does not matter if the window is open when the deserialization happens, or if it is opened afterwards.

If no window is able to swallow said region, then it will remain as an empty container.

Do notice that you will have to set up your desktop once, and then call the serialize command, as the previous layouts you may have created does not contain the necessary information for this system to work.

**Comment by crisidev** [comment]
I tested this out extensively and it works fine. It has some limitations when more windows of the same application (ie terminals) are on the same desktop since it relies on windows titles, but I don't think there is a lot we can do for it. 

Is it possible to merge this branch on master to have it available with normal installation channels? Or is there other work needed?

**Comment by koekeishiya** [comment]
It should be ready for merge to master, but I'm not yet sold on this system being better than the current system.

**Comment by crisidev** [comment]
Here's why I think it is better: if you have different applications on the same desktop and you deserialize, windows may be moved. Having at least the window type in the serialisation state allows different applications to stay where they were.

**Comment by koekeishiya** [comment]
I agree that is an improvement over the old system, but we lose the ability to "progress" the layout.


Currently, if you have a layout with 3 containers, and only 1 window open, the window takes the full screen. When you open a new window, that new window tiles into the layout, and so with the 3rd window.

The new system can't support that, and will just leave you with most of the screen empty, if there is only one window open. Now this could be the desired behaviour for all I know, as I don't use layouts much.

I looked into i3, and i3 basically has placeholders window instead of leaving the region completely empty, but doing that in macOS is a pain to put it lightly.

**Comment by crisidev** [comment]
I don't use layout much as well.. can you maybe make this configurable? So people can choose the behaviour via cvar?

**Comment by dominiklohmann** [comment]
I have tested the PR #390 thoroughly and came to a similar conclusion: It is clearly an improvement over the current system, but it also has its downsides, such as the mentioned inability to "progress the layout" and the added setup/configuration complexity.

Before merging this should be thoroughly documented I think.

**Comment by kerryj89** [comment]
From a usage point of view, I see how this behavior would be problematic for those who like working on the whim, creating new window instances on the fly and shifting windows about.  I get the feeling that was the author's intended audience and perhaps closer to his workflow, but I may be wrong.

There are those of us who work with a limited set of applications that would be totally fine with giving up some of the dynamicness in favor of a more territorial static setup.  It'll also help save us some of the hassle of having to move applications back into their spot if chunkwm can get the easy ones right for us.

How much more testing do we need until this gets merged in?

**Comment by koekeishiya** [comment]
> I get the feeling that was the author's intended audience and perhaps closer to his workflow, but I may be wrong.

This is pretty much my workflow, yes.

I don't think this really needs more testing, but if we are to support both the new and old method as briefly mentioned by @crisidev earlier, the code-structure would have to be changed to support that.

What I did in that branch was more or less just testing the idea. I think we could have both versions live side by side, and which one is active depends on some config-setting.
The layout serialization would need some additional metadata to track which system a given layout-file is supposed to work with.
